File: ctci/java/ch1-arrays-and-strings/q3_urlify/Question.java
package q3_urlify;

/**
 * Created by jjlee on 7/10/16.
 */
public class Question {
    /* iterate backwards using char array */
    public static void urlify(char[] s, int len) {
        int numSpaces = 0;
        for (int i = len - 1; i >= 0; i--) {
            if (s[i] == ' ')
                numSpaces++;
        }
        int urlIndex = len + 2 * numSpaces - 1;
        for (int i = len - 1; i >= 0; i--) {
            if (s[i] == ' ') {
                s[urlIndex--] = '0';
                s[urlIndex--] = '2';
                s[urlIndex--] = '%';
            } else {
                s[urlIndex--] = s[i];
            }
        }
    }
}

File: ctci/python/ch1-arrays-and-strings/q3-urlify/Question.py
# iterate backwards using char array
def urlify(s, length):
    num_spaces = s[:length].count(' ')
    url_index = length + 2 * num_spaces - 1
    for i in xrange(length - 1, -1, -1):
        if s[i] == ' ':
            s[url_index] = '0'
            s[url_index - 1] = '2'
            s[url_index - 2] = '%'
            url_index -= 3
        else:
            s[url_index] = s[i]
            url_index -= 1